KAFKA-15181: fix(storage): wait for consumer to be synced after assigning partitions

jeqo opened the pull request:
This fixes a race condition where cache is not ready but TBRLMM is marked as initialized anyway, causing replica to fail as remote log metadata is not found.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jeqo committed:
chore(storage): nit fixes on TBRLMM tests

mdedetrich commented:
Pinging @satishd and @junrao since its tiered storage related.

mdedetrich reviewed (approved):
Implementation looks good with me, also appreciate the minor code improvements. Only question is whether a test is required for this, and if so is there some prior art for testing `waitTillConsumptionCatchesUp`.

We are also going to test this implementation internally to verify there are no issues.

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -78,45 +88,42 @@ public void startConsumerThread() {
     /**
      * Waits if necessary for the consumption to reach the offset of the given {@code recordMetadata}.
      *
-     * @param recordMetadata record metadata to be checked for consumption.
```
Please remember to update the javadoc.

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -78,45 +88,42 @@ public void startConsumerThread() {
-     * @param recordMetadata record metadata to be checked for consumption.
-     * @param timeoutMs      wait timeout in milli seconds
+     * @param timeoutMs wait timeout in milliseconds
      * @throws TimeoutException if this method execution did not complete with in the given {@code timeoutMs}.
```
javadoc needs to be updated.

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -78,45 +88,42 @@ public void startConsumerThread() {
         // If the current assignment does not have the subscription for this partition then return immediately.
         if (!consumerTask.isPartitionAssigned(partition)) {
-            throw new KafkaException("This consumer is not subscribed to the target partition " + partition + " on which message is produced.");
+            throw new KafkaException("This consumer is not subscribed to the target partition " + partition + " assigned. " +
+                    "Partitions currently assigned: " + consumerTask.partitionsAssigned());
```
Thanks for adding more context for the error message. But I don't understand why this is added:
`This consumer is not subscribed to the target partition " + partition + [" assigned. "]` <-- why?

jeqo replied:
Sure, let me rephrase that to 
```
            throw new KafkaException("This consumer is not assigned to the target partition " + partition + ". " +
                    "Partitions currently assigned: " + consumerTask.metadataPartitionsAssigned());

```
as partitions are not technically subscribed, but manually assigned.

showuon commented on storage/src/test/resources/log4j.properties:
```diff
@@ -19,4 +19,4 @@ log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
 
 log4j.logger.org.apache.kafka.server.log.remote.storage=INFO
-log4j.logger.org.apache.kafka.server.log.remote.metadata.storage=INFO
+log4j.logger.org.apache.kafka.server.log.remote.metadata.storage=DEBUG
```
I don't we should default enable `DEBUG` log level. Should we revert this change?

jeqo replied:
Enabled it because it was hard to troubleshoot test issues without it; also only enabled on tests. Though I'm ok reverting it.

mdedetrich replied:
Agreed, we shouldn't default to `DEBUG` for production code otherwise it produces way too many logs.

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -139,6 +147,33 @@ public void close() throws IOException {
+     * Whenever there is a change on the meta partitions assgined to the TBRLMM,
+     * wait for consumer (therefore the cache as well) to be in-sync before continuing
+     */
+    private void waitTillMetaPartitionIsInSync() {
```
Could we change the term `inSync` to `catchUp` as above?

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -139,6 +147,33 @@ public void close() throws IOException {
+                }
+            }
+        } catch (ExecutionException | InterruptedException | TimeoutException e) {
+            log.error("Error encountered while consuming from meta partitions", e);
```
I think we should output the partition name, and remove the ambiguous "meta partition", ex:
`("Error encountered while consuming from remote log metadata partitions: {}", consumerTask.partitionsAssigned(), e)`

showuon replied:
Should we throw exception here?

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -363,7 +363,7 @@ private void initializeResources() {
         long startTimeMs = time.milliseconds();
         AdminClient adminClient = null;
         try {
-            adminClient = AdminClient.create(rlmmConfig.producerProperties());
+            adminClient = AdminClient.create(rlmmConfig.commonProperties());
```
Thanks for the improvement. Could we use try-with-resource here?

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -139,6 +147,33 @@ public void close() throws IOException {
+    }
+
+    /**
+     * Whenever there is a change on the meta partitions assgined to the TBRLMM,
```
Could we make this clear? ex:
`Whenever there is a change on the remote log metadata topic partitions assgined to the topic based remote log metadata manager,`

showuon reviewed (commented):
@jeqo , nice find and fix! Left some comments.

showuon reviewed (commented):
One more comment, I don't see tests for the new added method. Could you add some tests for them?

jeqo committed:
feat(storage): add common set of properties

TBRLMM needs to create admin client at separate instances, and common set of configs already existing are enough -- no need to reuse producer/consumer configs

jeqo committed:
test(storage): add test case to validate cache is in-sync on assignments

When partitions are assigned on leadership changes cache should be ready to reply, otherwise there is a risk of reads on cache without partitions assigned, and failing.

This can happen on bootstrap where partitions are assigned, then added to partitions pending to load, and then taking some time to load data from partitions.

The waiting happening on this test should then be actual part of the logic when assigning partitions to TBRLMM, and give the same guarantees as when writing and then waiting for data to be consumed.

jeqo committed:
fix(storage): wait for consumer to be synced after assigning partitions

This fixes a race condition where cache is not ready but TBRLMM is marked as initialized anyway, causing replica to fail as remote log metadata is not found.

jeqo commented:
@showuon thanks for your feedback! I have applied most suggestions -- let me know if I'm missing anything.

About the test, I followed @mdedetrich advise on piggybacking on existing testing of `waitTillConsumptionCatchesUp` and actually removing the waiting from `TopicBasedRemoteLogMetadataManagerTest` to prove the new behavior when assigning partitions. 

I tried to add some scenarios to `TopicBasedRemoteLogMetadataManagerRestartTest` but the current threads initialization haven't been able to reproduce. Maybe something to improve in a follow up PR, I think this one may be related: https://github.com/apache/kafka/pull/13689

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -76,47 +94,50 @@ public void startConsumerThread() {
-                                             long timeoutMs) throws TimeoutException {
-        final int partition = recordMetadata.partition();
+    public void waitTillConsumptionCatchesUp(int partition, long offset, long timeoutMs) throws TimeoutException {
         final long consumeCheckIntervalMs = Math.min(CONSUME_RECHECK_INTERVAL_MS, timeoutMs);
```
I think we should log something as INFO when waiting the consumer catching up, ex:
`log.info("waiting until consumer is caught up with the target partition {}", partition);`
so that we can know why it's pending without enabling debug log. WDYT?

mdedetrich replied:
This makes sense especially since we are dealing with remote storage and hence we may get high lag times.

showuon reviewed (commented):
LGTM, just a minor comment.

> About the test, I followed @mdedetrich advise on piggybacking on existing testing of waitTillConsumptionCatchesUp and actually removing the waiting from TopicBasedRemoteLogMetadataManagerTest to prove the new behavior when assigning partitions.

Fair enough!

jeqo committed:
refactor: info log when waiting for consumer to catch up

showuon reviewed (approved):
LGTM! Thanks for the fix!

kamalcph commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -209,7 +214,7 @@ private void maybeSyncCommittedDataAndOffsets(boolean forceSync) {
                         remotePartitionMetadataEventHandler.syncLogMetadataSnapshot(topicIdPartition, metadataPartition, offset);
                     } else {
-                        log.debug("Skipping syncup of the remote-log-metadata-file for partition:{} , with remote log metadata partition{}, and no offset",
+                        log.debug("Skipping sync-up of the remote-log-metadata-file for partition:{} , with remote log metadata partition{}, and no offset",
```
partition{} -> partition {}

kamalcph reviewed (commented):
@jeqo Thanks for the PR!

1. Can we split this patch into two?  a) Refactoring the existing code and b) Specific changes that address KAFKA-15181 ticket.
2. We are instantiating `AdminClient` (one more client) to do listOffset calls in ConsumerManager to ensure that once the user-partitions are assigned, all the messages are caught-up from the respective metadata-partition. 

Both Consumer and Admin clients, can do list-offset calls. Instead, of adding one more Admin client, can we move the initialisation logic to RemoteLogMetadataCache.java class by introducing a latch. All the calls to read/write from remote will hit the RemoteLogMetadataCache and the `ConsumerTask` can mark the partition as initialised once it reads all the messages.

kamalcph commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -139,6 +163,35 @@ public void close() throws IOException {
 
     public void addAssignmentsForPartitions(Set<TopicIdPartition> partitions) {
         consumerTask.addAssignmentsForPartitions(partitions);
+        waitTillMetaPartitionConsumptionCatchesUp();
```
This is a blocking call. It won't return until the partition is ready/consumption-catches-up. In high throughput cluster, this can anywhere take order of seconds to complete. 

Moreover, when you reassign the partitions, the controller will send the LISR request for one partition at a time and the LISR request handling path shouldn't have any blocking call so that the controller can send the next LISR request on time. Otherwise, LISR request gets timed out.

showuon replied:
Good point @kamalcph ! Do you have any suggestion here?

mdedetrich replied:
I am also unclear as to what we should do here, I mean we could return a `CompletableFuture<Void>` but this would involve changing the interface?

kamalcph replied:
One approach we can take is to wait for metadata partition to catch up asynchronously. All the calls to read the remote log will hit the [RemotePartitionMetadataStore#getRemoteLogMetadataCache](https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java#L147), there we can ensure that the partition is initialised via a latch. 

The [ConsumerTask](https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java) can keep a marker of the latest offset of metadata-partition when user-partitions are assigned to them and can mark that user-partition as initialised in the cache once it read/re-read messages up-to that point. 

We can have a latch with timeout of 2 mins to avoid holding the caller thread for a long amount of time and throw a retry-able error.

showuon replied:
Sounds great! I like the idea that we lazily wait for the consumer to catch up. One comment is why don't we return retriable error immediately, and ask the client to try it again, like when the leader is under election, the client will get `NOT_LEADER_OR_FOLLOWER` error immediately and retry, instead of waiting in the broker side until leader election completed. WDYT?

kamalcph replied:
That also works! We have to look around for all the callers:

1. Replica fetcher thread will retry the request after the `replica.fetch.backoff.ms` once the partition is marked as error/failed.
2. Consumers may resend the FETCH request as soon as it fails without the request quota. 

We can also think about adding one more API in [RemoteLogMetadataManager](https://github.com/apache/kafka/blob/trunk/storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataManager.java) to ensure that the partition is initialised before copying the segments (or) when the replica thread is being initialised. 

```
boolean isInitialized(TopicIdPartition topicIdPartition) throws RemoteStorageException;
```

satishd replied:
The latching mechanism as mentioned by @kamalcph is what we are currently using internally. @abhijeetk88 is planning to pull those changes to trunk and raise PR this week to handle the scenario.

satishd reviewed (commented):
Thanks @jeqo for the PR. Let us keep the changes separate as @kamalcph suggested in other [comment](https://github.com/apache/kafka/pull/14012#pullrequestreview-1533335715). We should avoid blocking the calls while registering a partition to RLMM, which is handled using latches in internal branch, left a [comment](https://github.com/apache/kafka/pull/14012#discussion_r1266471753) on the same. Let us have this PR with the cleanup and other minor changes.

jeqo committed:
fixup! test(storage): add test case to validate cache is in-sync on assignments

jeqo commented:
Thanks @satishd and @kamalcph, I've moved refactoring changes to https://github.com/apache/kafka/pull/14045, please take a look.

About the fix, looking forward to take a look to your PR. Will hold changes on this PR then, many thanks.

satishd commented:
Thanks @jeqo for moving the changes to https://github.com/apache/kafka/pull/14045. As I already merged that PR, closing this PR.
